### PR TITLE
Lesson9

### PR DESCRIPTION
**Pull request by mhope21**
This did not have quite enough explanations. The tutorials in the resource materials were hard to complete because of version mismatches with the dependencies. I hope that is the right word. I mean mismatches between rails, capybara, rspec,, and shoulda. Are those called dependencies?

### CODE REVIEW

File: spec/models/customer_spec.rb [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  it "is not valid if the phone number is not 10 chars" do
+    subject.phone = "12345"  # Less than 10 digits
+    expect(subject).to_not be_valid
+
+    subject.phone = "12345678901"  # More than 10 digits
+    expect(subject).to_not be_valid
+
+    subject.phone = "1234567890"  # Exactly 10 digits
+    expect(subject).to be_valid
+  end
```
Comment (dsinn) [inline spec/models/customer_spec.rb]: It's good to be thorough with your tests, but when there are multiple assertions in a row -- especially when they're all independent of each other -- we typically separate them into separate scenarios. (If you wish to group multiple scenarios together, you can do so with contexts as shown in [RSpec's examples](https://rspec.info/features/3-13/rspec-core/example-groups/basic-structure/))

The problem with multiple assertions in a row is that as soon as one fails, the scenario stops and you won't know what would've happened with the rest of the assertions. So in this case, if a bug caused `12345` to be valid, you wouldn't know whether `12345678901` or `1234567890` would pass or fail, and that is information that you'd want from the tests.

In fact, the helpful comments you added are a further indication that it's worth splitting into multiple scenarios.

```suggestion
  it "is not valid if the phone number has less than 10 digits" do
    subject.phone = "12345"
    expect(subject).to_not be_valid
  end
  
  it "is not valid if the phone number has more than 10 digits"
    subject.phone = "12345678901"
    expect(subject).to_not be_valid
  end
  
  it "is valid if the phone number has exactly 10 digits"
    subject.phone = "1234567890"
    expect(subject).to be_valid
  end
```

File: spec/requests/customers_spec.rb [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    it "updates an entry and redirects to the show path for the customer" do
+      customer = FactoryBot.create(:customer)
+      # Attempt to update customer with valid data
+      put customer_path(customer), params: {customer: {phone: "1234567890"}}
+      customer.reload
+      # Check that valid data was saved
+      expect(customer.phone).to eq("1234567890")
+
+      expect(response).to redirect_to(customer_path(customer))
+      expect(response).to have_http_status(:redirect)
```
Comment (dsinn) [inline spec/requests/customers_spec.rb:63]: I'd expect the second assertion to imply the first, so I don't think you need both.

```suggestion
      expect(response).to redirect_to(customer_path(customer))
```

**Review by dsinn — APPROVED** [review]
> This did not have quite enough explanations. The tutorials in the resource materials were hard to complete because of version mismatches with the dependencies. I hope that is the right word. I mean mismatches between rails, capybara, rspec,, and shoulda. Are those called dependencies?

Sorry to hear about the difficulty in this lesson, and thank you for the feedback. Yes, these are dependencies. I'll submit it for John to see.

Personally, I think it would be a relatively easy fix to constrain the versions to what the resource materials use via the "version specifiers" described [here](https://bundler.io/guides/gemfile.html#:~:text=Most%20of%20the-,version%20specifiers,-%2C%20like%20%3E%3D%201.0%2C%20are).

**Comment by mhope21** [comment]
I appreciate your thoughts.  Thank you for helping me to improve. I will include your suggestions.

**Pull request merged**